docs/spec/relayer/Chain.tla

```diff
@@ -111,29 +183,98 @@ UpdateChainStore(chainID, datagrams) ==
 AdvanceChain ==
     /\ chainStore.height + 1 \in Heights
     /\ chainStore' = [chainStore EXCEPT !.height = chainStore.height + 1]
-    /\ UNCHANGED <<incomingDatagrams, history>>
+    /\ UNCHANGED <<incomingDatagrams, incomingPacketDatagrams, history>>
+    /\ UNCHANGED <<packetLog, appPacketSeq>>
+
+\* Send a packet
+SendPacket ==   
+    \* Create a packet 
+    \* Abstract away from packet data, ports, and timestamp. 
+    \* Assume timeoutHeight is MaxHeight + 1  
+    LET packet == AsPacket([
+        sequence |-> appPacketSeq,
+        timeoutHeight |-> MaxHeight + 1,
+        srcChannelID |-> GetChannelID(ChainID),
+        dstChannelID |-> GetChannelID(GetCounterpartyChainID(ChainID))
+    ]) IN
+    LET connectionEnd == chainStore.connectionEnd IN
+    LET channelEnd == chainStore.connectionEnd.channelEnd IN
+    LET latestClientHeight == GetMaxCounterpartyClientHeight(chainStore) IN
+    \* channel end is neither null nor closed
+    /\ channelEnd.state \notin {"UNINIT", "CLOSED"}
+    \* connection end is initialized
+    /\ connectionEnd.state /= "UNINIT"
+    \* timeout height has not passed
+    /\ \/ packet.timeoutHeight = 0 
+       \/ latestClientHeight < packet.timeoutHeight
+    /\ \* if the channel is ordered, check if packetSeq is nextSendSeq, 
+       \* add a packet committment in the chain store, and increase nextSendSeq
+       \/ /\ channelEnd.order = "ORDERED"
+          /\ packet.sequence = channelEnd.nextSendSeq
+          /\ packet.sequence <= MaxPacketSeq
```

Comment:
This is a model constraint, not the handler one; ideally they should not be mixed and model constraint can be a precondition at the higher level.

Reply:
I've refactored this action a bit.